Allow apps to receive and remove data being used by web extensions

kiaraarose opened the pull request:
#### cb7ebb2b9986cb6461ec33039399c3ed4795a61e
<pre>
Allow apps to receive and remove data being used by web extensions
<a href="https://bugs.webkit.org/show_bug.cgi?id=269112">https://bugs.webkit.org/show_bug.cgi?id=269112</a>
<a href="https://rdar.apple.com/122677362">rdar://122677362</a>

Reviewed by Timothy Hatcher.

This patch allow apps to receive _WKWebExtensionDataRecords providing information on:
- How much data an extension is using
- What type of data (local, sync, session) is being used

In addition, apps can specify data records to remove. This patch also renames WebExtensionStorageType
to WebExtensionDataType to have the naming be consistent with the new APIs being added.

* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/Extensions/WebExtensionDataType.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
(WebKit::toAPIPrefixString):
* Source/WebKit/Shared/Extensions/WebExtensionStorage.serialization.in:

* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm:
(-[_WKWebExtensionSQLiteStore close]):
This method may be called on the _databaseQueue. To prevent a deadlock caused by the dispatch_sync,
check if we&apos;re on the main thread. If we are, we&apos;re safe to dispatch_sync on the _databaseQueue.
If we&apos;re not, we&apos;re already on the _databaseQueue, so no need to dispatch the call.

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(+[_WKWebExtensionController allExtensionDataTypes]):
(-[_WKWebExtensionController fetchDataRecordsOfTypes:completionHandler:]):
(-[_WKWebExtensionController fetchDataRecordOfTypes:forExtensionContext:completionHandler:]):
(-[_WKWebExtensionController removeDataOfTypes:forDataRecords:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.h:
Copied from Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.h.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.mm: Added.
(-[_WKWebExtensionDataRecord dealloc]):
(-[_WKWebExtensionDataRecord isEqual:]):
(-[_WKWebExtensionDataRecord displayName]):
(-[_WKWebExtensionDataRecord uniqueIdentifier]):
(-[_WKWebExtensionDataRecord dataTypes]):
(-[_WKWebExtensionDataRecord totalSize]):
(-[_WKWebExtensionDataRecord sizeOfDataTypes:]):
(-[_WKWebExtensionDataRecord _apiObject]):
(-[_WKWebExtensionDataRecord _webExtensionDataRecord]):
Helper methods to convert between data types.
(WebKit::makeVectorElement):
(WebKit::makeNSArrayElement):
(WebKit::toWebExtensionDataRecords):
(WebKit::toAPI):

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecordInternal.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecordPrivate.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataType.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataType.mm:
Helper methods to convert between data types.
(WebKit::toWebExtensionDataTypes):
(WebKit::toAPI):

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataTypeInternal.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
(WebKit::WebExtensionContext::storageGet):
(WebKit::WebExtensionContext::storageGetBytesInUse):
(WebKit::WebExtensionContext::storageSet):
(WebKit::WebExtensionContext::storageRemove):
(WebKit::WebExtensionContext::storageClear):
(WebKit::WebExtensionContext::storageSetAccessLevel):
(WebKit::WebExtensionContext::fireStorageChangedEventIfNeeded):
Switch from WebExtensionStorageType to WebExtensionDataType.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):

(WebKit::WebExtensionContext::readStateFromPath):
(WebKit::WebExtensionContext::readDisplayNameAndLastBaseURLFromState):
(WebKit::WebExtensionContext::readLastBaseURLFromState):
Static helper methods to read last seen display name and baseURL from disk.

(WebKit::WebExtensionContext::readStateFromStorage):
(WebKit::WebExtensionContext::quoataForStorageType):
(WebKit::WebExtensionContext::localStorageStore):
(WebKit::WebExtensionContext::sessionStorageStore):
(WebKit::WebExtensionContext::syncStorageStore):
Switch from WebExtensionStorageType to WebExtensionDataType.
(WebKit::WebExtensionContext::storageForType):
Make public so it can be used in by the WebExtensionController.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::getDataRecords):
(WebKit::WebExtensionController::getDataRecord):
(WebKit::WebExtensionController::removeData):
(WebKit::WebExtensionController::calculateStorageSize):
(WebKit::WebExtensionController::removeStorage):
(WebKit::WebExtensionController::stateFilePath const):
(WebKit::WebExtensionController::storageDirectory const):
Use the FileSystem to get the storage directory for an extension with a given uniqueIdentifier.
(WebKit::WebExtensionController::sqliteStore):
Helper method to return the sql store for a given data type. If the type is WebExtensionDataType::Session,
uses the WebExtensionContext to check if the context is loaded. If it&apos;s loaded, use the sessionStorageStore
on the context to get the data, since creating a new instance of the session storage object will not have the
correct info.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
(-[_WKWebExtensionStorageSQLiteStore initWithUniqueIdentifier:storageType:directory:usesInMemoryDatabase:]):
(-[_WKWebExtensionStorageSQLiteStore _databaseURL]):
Switch from WebExtensionStorageType to WebExtensionDataType.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::plistFileName):
Static method returning the file name for the extension plist.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionDataRecord.cpp:
Renamed from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
(WebKit::WebExtensionDataRecord::WebExtensionDataRecord):
(WebKit::WebExtensionDataRecord::operator== const):
(WebKit::WebExtensionDataRecord::totalSize const):
(WebKit::WebExtensionDataRecord::sizeOfTypes const):
(WebKit::WebExtensionDataRecord::types const):

* Source/WebKit/UIProcess/Extensions/WebExtensionDataRecord.h: Added.
(WebKit::WebExtensionDataRecord::create):
(WebKit::WebExtensionDataRecord::displayName const):
(WebKit::WebExtensionDataRecord::uniqueIdentifier const):
(WebKit::WebExtensionDataRecord::sizeOfType const):
(WebKit::WebExtensionDataRecord::setSizeOfType):
(WebKit::WebExtensionDataRecord::wrapper const):
(WebKit::WebExtensionDataRecordHolder::create):
(WebKit::WebExtensionDataRecordHolder::WebExtensionDataRecordHolder):
Ref counted class to ensure the completion handler doesn&apos;t get called too early
when calculating storage size.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::isPropertyAllowed):
(WebKit::WebExtensionAPIStorageArea::quotaBytes):
(WebKit::WebExtensionAPIStorageArea::quotaBytesPerItem):
(WebKit::WebExtensionAPIStorageArea::maxItems):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerHour):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerMinute):
Switch from WebExtensionStorageType to WebExtensionDataType.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm:
(WebKit::WebExtensionAPIStorage::local):
(WebKit::WebExtensionAPIStorage::session):
(WebKit::WebExtensionAPIStorage::sync):
(WebKit::WebExtensionAPIStorage::storageAreaForType):
(WebKit::WebExtensionContextProxy::dispatchStorageChangedEvent):
Switch from WebExtensionStorageType to WebExtensionDataType.

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
(WebKit::WebExtensionAPIStorageArea::WebExtensionAPIStorageArea):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
Switch from WebExtensionStorageType to WebExtensionDataType.

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionDataRecord.mm: Added.
(TestWebKitAPI::TEST):
The test for removing the data records are disabled because they were too
flaky on the bots.
<a href="https://bugs.webkit.org/show_bug.cgi?id=269329">https://bugs.webkit.org/show_bug.cgi?id=269329</a> is tracking enabling them.

Canonical link: <a href="https://commits.webkit.org/274603@main">https://commits.webkit.org/274603@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8822abbb742aa040571562968cbab14b08bef5d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39540 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41894 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42074 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35440 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41846 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21401 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15848 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40114 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15619 "Found 60 new test failures: accessibility/changing-aria-hidden-with-display-none-parent.html, accessibility/input-time-label-value.html, compositing/overflow/nested-scrolling.html, compositing/repaint/copy-forward-dirty-region-purged-backbuffer.html, compositing/repaint/copy-forward-dirty-region-purged.html, compositing/repaint/needs-no-display-volatile-repaint.html, compositing/repaint/needs-no-display-volatile.html, crypto/subtle/aes-cbc-cfb-decrypt-malformed-parameters.html, crypto/subtle/aes-cbc-cfb-encrypt-malformed-parameters.html, crypto/subtle/aes-cfb-generate-export-key-jwk-length-128.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34230 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13546 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/39407 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13521 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35181 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43352 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35893 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35512 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39308 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14352 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11826 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15958 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16006 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5187 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/Platform/cocoa/CocoaHelpers.h:
```diff
@@ -121,6 +121,7 @@ NSString *escapeCharactersInString(NSString *, NSString *charactersToEscape);
 NSDate *toAPI(const WallTime&);
 WallTime toImpl(NSDate *);
 
+
```
remove

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -80,6 +80,7 @@
 #import <WebCore/UserScript.h>
 #import <pal/spi/cocoa/NSKeyedUnarchiverSPI.h>
 #import <wtf/BlockPtr.h>
+#import <wtf/CallbackAggregator.h>
```
will remove

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionDataType.h:
```diff
@@ -0,0 +1,70 @@
+} // namespace WebKit
+
+namespace WTF {
+template<> struct DefaultHash<WebKit::WebExtensionDataType> : IntHash<WebKit::WebExtensionDataType> { };
```
Add newline above.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
```diff
@@ -122,6 +124,34 @@ WK_CLASS_AVAILABLE(macos(13.3), ios(16.4))
+  @param completionHandler A block to invoke when the data record has been fetched.
+  @note The extension does not need to be loaded to be included in the result.
+*/
+- (void)fetchDataRecordOfTypes:(NSSet<_WKWebExtensionDataType> *)dataTypes forExtensionContext:(_WKWebExtensionContext *)extensionContext completionHandler:(void (^)(_WKWebExtensionDataRecord * _Nullable))completionHandler WK_SWIFT_ASYNC_NAME(dataRecord(ofTypes:forExtensionContext:));
```
Likely should be: `WK_SWIFT_ASYNC_NAME(dataRecord(ofTypes:for:))`

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.mm:
```diff
@@ -0,0 +1,152 @@
+
+NSArray *toAPI(const Vector<Ref<WebExtensionDataRecord>>& records)
+{
+    return createNSArray(records).autorelease();
```
`createNSArray(records).get()` should be fine, since this is ARC code.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecordInternal.h:
```diff
@@ -0,0 +1,56 @@
+
+@end
+
+namespace WebKit {
+
+Vector<Ref<WebExtensionDataRecord>> toWebExtensionDataRecords(NSArray *);
+NSArray *toAPI(const Vector<Ref<WebExtensionDataRecord>>&);
+
+} // namespace WebKit
```
I like putting these here! We should do that for the other helpers we have like this.

xeenon commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataType.h:
```diff
@@ -0,0 +1,45 @@
+typedef NSString * _WKWebExtensionDataType NS_TYPED_EXTENSIBLE_ENUM NS_SWIFT_NAME(_WKWebExtension.DataType);
+
+// FIXME: KIARA: remove window and indexedDB
+/*! @abstract Specifies local storage, including `browser.storage.local`, `window.localStorage`, and IndexedDB databases. */
```
FYI

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -65,6 +69,122 @@
+    auto recordHolder = RecordHolder::create();
+    auto aggregator = MainRunLoopCallbackAggregator::create([recordHolder, completionHandler = WTFMove(completionHandler)]() mutable {
+        Vector<Ref<WebExtensionDataRecord>> records;
+        for (auto entry : recordHolder->recordsMap)
```
`auto&`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -65,6 +69,122 @@
+    if (!matchingUniqueIdentifier) {
+        completionHandler(nullptr);
+        return;
+    }
```
This still assumes the extension context is loaded. While this works, I think we can make this work for unloaded contexts too. The passed in context does not need to be loaded, since we just need to ask for its unique identifier.

```c++
auto uniqueIdentifier = extensionContext.uniqueIdentifier();
auto uniqueIdentifiers = FileSystem::listDirectory(m_configuration->storageDirectory());
if (!uniqueIdentifiers.contains(uniqueIdentifier)) {
    completionHandler(nullptr);
    return;
}

auto storageDirectory = FileSystem::pathByAppendingComponent(m_configuration->storageDirectory(), uniqueIdentifier);
auto stateFilePath = FileSystem::pathByAppendingComponent(storageDirectory, "State.plist"_s);

String displayName;
URL lastBaseURL;
if (!WebExtensionContext::readDisplayNameAndLastBaseURLFromState(stateFilePath, displayName, lastBaseURL)) {
    completionHandler(nullptr);
    return;
}
```

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -65,6 +69,122 @@
+    auto uniqueIdentifiers = FileSystem::listDirectory(m_configuration->storageDirectory());
+    for (auto& uniqueIdentifier : uniqueIdentifiers) {
+        auto storageDirectory = FileSystem::pathByAppendingComponent(m_configuration->storageDirectory(), uniqueIdentifier);
+        auto stateFilePath = FileSystem::pathByAppendingComponent(storageDirectory, "State.plist"_s);
```
```c++
"State.plist"_s
```
This should likely be in a static getter method on `WebExtensionContext` so it does not expose the file name here (and below) and we can change it in `WebExtensionContext` freely if needed.

You can make a private helper method on `WebExtensionController` to get the state path in one shot for a unique identifier, so you don't have to do this in a couple places in this patch.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -65,6 +69,122 @@
+
+    auto uniqueIdentifiers = FileSystem::listDirectory(m_configuration->storageDirectory());
+    for (auto& uniqueIdentifier : uniqueIdentifiers) {
+        auto storageDirectory = FileSystem::pathByAppendingComponent(m_configuration->storageDirectory(), uniqueIdentifier);
```
This can be a private helper method that is used here (and elsewhere in this file).

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -65,6 +69,122 @@
+
+        for (auto dataType : record.get().types()) {
+            // FIXME: <https://webkit.org/b/269100> Remove window.localStorage, window.sessionStorage, indexedDB.
+            auto *storage = [[_WKWebExtensionStorageSQLiteStore alloc] initWithUniqueIdentifier:uniqueIdentifier storageType:dataType directory:storageDirectory usesInMemoryDatabase:NO];
```
This could be a helper method that returns `_WKWebExtensionStorageSQLiteStore *`, which also can be used below in `calculateStorageSizeForDataType()`.

This should also continue early, or have the helper return `nil` if `dataType` is not `Local` or `Sync`, until the FIXME is fixed. Since `usesInMemoryDatabase:NO` will be wrong for `Session`.

kiaraarose replied:
yea passing `usesInMemoryDatabase:NO` is wrong here. fixed this to pass in `dataType::WebExtensionDataTypeSession` for `usesInMemoryDatabase`

xeenon replied:
@kiaraarose But we will need to use the right session DB, from the loaded `WebExtensionContext`, since a new one will just be empty. So it is likely best to not do anything when type is Session for now.

kiaraarose replied:
True. But since storage used for Session will be cleared when a context unloads, I'll just add a check to see if the context is loaded and then use the `sessionStorageStore()` on the WebExtensionContext to handle calculating/removing storage. Thanks for pointing that out!

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -65,6 +69,122 @@
+
+void WebExtensionController::calculateStorageSizeForDataType(WebExtensionDataType dataType, const String& storageDirectory, const String& uniqueIdentifier, CompletionHandler<void(size_t)>&& completionHandler)
+{
+    auto *storage = [[_WKWebExtensionStorageSQLiteStore alloc] initWithUniqueIdentifier:uniqueIdentifier storageType:dataType directory:storageDirectory usesInMemoryDatabase:NO];
```
Could use the helper here.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionDataRecord.h:
```diff
@@ -0,0 +1,95 @@
+    HashMap<Type, size_t> m_typeSizes;
+};
+
+class RecordHolder : public RefCounted<RecordHolder> {
```
`WebExtensionDataRecordHolder`

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -240,6 +242,8 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     auto lastSeenBaseURL = URL { objectForKey<NSString>(m_state, lastSeenBaseURLStateKey) };
     [m_state setObject:(NSString *)m_baseURL.string() forKey:lastSeenBaseURLStateKey];
 
+    [m_state setObject:extension().displayName() forKey:lastSeenDisplayNameStateKey];
```
this is what's causing the test failures on the PR. switch to
```
if (NSString *displayName = extension().displayName())
    [m_state setObject:displayName forKey:lastSeenDisplayNameStateKey];
```

kiaraarose committed:
Allow apps to receive and remove data being used by web extensions
https://bugs.webkit.org/show_bug.cgi?id=269112
rdar://122677362

Reviewed by Timothy Hatcher.

This patch allow apps to receive _WKWebExtensionDataRecords providing information on:
- How much data an extension is using
- What type of data (local, sync, session) is being used

In addition, apps can specify data records to remove. This patch also renames WebExtensionStorageType
to WebExtensionDataType to have the naming be consistent with the new APIs being added.

* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/Platform/cocoa/CocoaHelpers.h:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/Extensions/WebExtensionDataType.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
(WebKit::toAPIPrefixString):
* Source/WebKit/Shared/Extensions/WebExtensionStorage.serialization.in:

* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm:
(-[_WKWebExtensionSQLiteStore close]):
This method may be called on the _databaseQueue. To prevent a deadlock caused by the dispatch_sync,
check if we're on the main thread. If we are, we're safe to dispatch_sync on the _databaseQueue.
If we're not, we're already on the _databaseQueue, so no need to dispatch the call.

* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(+[_WKWebExtensionController allExtensionDataTypes]):
(-[_WKWebExtensionController fetchDataRecordsOfTypes:completionHandler:]):
(-[_WKWebExtensionController fetchDataRecordOfTypes:forExtensionContext:completionHandler:]):
(-[_WKWebExtensionController removeDataOfTypes:forDataRecords:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.h:
Copied from Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.h.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecord.mm: Added.
(-[_WKWebExtensionDataRecord dealloc]):
(-[_WKWebExtensionDataRecord isEqual:]):
(-[_WKWebExtensionDataRecord displayName]):
(-[_WKWebExtensionDataRecord uniqueIdentifier]):
(-[_WKWebExtensionDataRecord dataTypes]):
(-[_WKWebExtensionDataRecord totalSize]):
(-[_WKWebExtensionDataRecord sizeOfDataTypes:]):
(-[_WKWebExtensionDataRecord _apiObject]):
(-[_WKWebExtensionDataRecord _webExtensionDataRecord]):
Helper methods to convert between data types.
(WebKit::makeVectorElement):
(WebKit::makeNSArrayElement):
(WebKit::toWebExtensionDataRecords):
(WebKit::toAPI):

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecordInternal.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataRecordPrivate.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataType.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataType.mm:
Helper methods to convert between data types.
(WebKit::toWebExtensionDataTypes):
(WebKit::toAPI):

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionDataTypeInternal.h:
Copied from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
(WebKit::WebExtensionContext::storageGet):
(WebKit::WebExtensionContext::storageGetBytesInUse):
(WebKit::WebExtensionContext::storageSet):
(WebKit::WebExtensionContext::storageRemove):
(WebKit::WebExtensionContext::storageClear):
(WebKit::WebExtensionContext::storageSetAccessLevel):
(WebKit::WebExtensionContext::fireStorageChangedEventIfNeeded):
Switch from WebExtensionStorageType to WebExtensionDataType.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::load):

(WebKit::WebExtensionContext::readStateFromPath):
(WebKit::WebExtensionContext::readDisplayNameAndLastBaseURLFromState):
(WebKit::WebExtensionContext::readLastBaseURLFromState):
Static helper methods to read last seen display name and baseURL from disk.

(WebKit::WebExtensionContext::readStateFromStorage):
(WebKit::WebExtensionContext::quoataForStorageType):
(WebKit::WebExtensionContext::localStorageStore):
(WebKit::WebExtensionContext::sessionStorageStore):
(WebKit::WebExtensionContext::syncStorageStore):
Switch from WebExtensionStorageType to WebExtensionDataType.
(WebKit::WebExtensionContext::storageForType):
Make public so it can be used in by the WebExtensionController.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::getDataRecords):
(WebKit::WebExtensionController::getDataRecord):
(WebKit::WebExtensionController::removeData):
(WebKit::WebExtensionController::calculateStorageSize):
(WebKit::WebExtensionController::removeStorage):
(WebKit::WebExtensionController::stateFilePath const):
(WebKit::WebExtensionController::storageDirectory const):
Use the FileSystem to get the storage directory for an extension with a given uniqueIdentifier.
(WebKit::WebExtensionController::sqliteStore):
Helper method to return the sql store for a given data type. If the type is WebExtensionDataType::Session,
uses the WebExtensionContext to check if the context is loaded. If it's loaded, use the sessionStorageStore
on the context to get the data, since creating a new instance of the session storage object will not have the
correct info.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
(-[_WKWebExtensionStorageSQLiteStore initWithUniqueIdentifier:storageType:directory:usesInMemoryDatabase:]):
(-[_WKWebExtensionStorageSQLiteStore _databaseURL]):
Switch from WebExtensionStorageType to WebExtensionDataType.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::plistFileName):
Static method returning the file name for the extension plist.

* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionDataRecord.cpp:
Renamed from Source/WebKit/Shared/Extensions/WebExtensionStorageType.h.
(WebKit::WebExtensionDataRecord::WebExtensionDataRecord):
(WebKit::WebExtensionDataRecord::operator== const):
(WebKit::WebExtensionDataRecord::totalSize const):
(WebKit::WebExtensionDataRecord::sizeOfTypes const):
(WebKit::WebExtensionDataRecord::types const):

* Source/WebKit/UIProcess/Extensions/WebExtensionDataRecord.h: Added.
(WebKit::WebExtensionDataRecord::create):
(WebKit::WebExtensionDataRecord::displayName const):
(WebKit::WebExtensionDataRecord::uniqueIdentifier const):
(WebKit::WebExtensionDataRecord::sizeOfType const):
(WebKit::WebExtensionDataRecord::setSizeOfType):
(WebKit::WebExtensionDataRecord::wrapper const):
(WebKit::WebExtensionDataRecordHolder::create):
(WebKit::WebExtensionDataRecordHolder::WebExtensionDataRecordHolder):
Ref counted class to ensure the completion handler doesn't get called too early
when calculating storage size.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::isPropertyAllowed):
(WebKit::WebExtensionAPIStorageArea::quotaBytes):
(WebKit::WebExtensionAPIStorageArea::quotaBytesPerItem):
(WebKit::WebExtensionAPIStorageArea::maxItems):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerHour):
(WebKit::WebExtensionAPIStorageArea::maxWriteOperationsPerMinute):
Switch from WebExtensionStorageType to WebExtensionDataType.

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageCocoa.mm:
(WebKit::WebExtensionAPIStorage::local):
(WebKit::WebExtensionAPIStorage::session):
(WebKit::WebExtensionAPIStorage::sync):
(WebKit::WebExtensionAPIStorage::storageAreaForType):
(WebKit::WebExtensionContextProxy::dispatchStorageChangedEvent):
Switch from WebExtensionStorageType to WebExtensionDataType.

* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorage.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
(WebKit::WebExtensionAPIStorageArea::WebExtensionAPIStorageArea):
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
Switch from WebExtensionStorageType to WebExtensionDataType.

* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionDataRecord.mm: Added.
(TestWebKitAPI::TEST):
The test for removing the data records are disabled because they were too
flaky on the bots.
https://bugs.webkit.org/show_bug.cgi?id=269329 is tracking enabling them.

Canonical link: https://commits.webkit.org/274603@main
